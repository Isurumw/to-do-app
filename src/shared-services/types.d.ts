declare interface IStorage {
  get(key: string): Promise<string | undefined>;
  getObject<P>(key: string): Promise<P | undefined>;
  set(key: string, value: string): Promise<void>;
  remove(key: string): void;
}

declare interface INetworkManager {
  request(
    route: string,
    method: 'POST' | 'GET' | 'PUT' | 'DELETE' | 'PATCH',
    variables?: object | null,
  ): any;
}
