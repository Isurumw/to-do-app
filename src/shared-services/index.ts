import Config from 'react-native-config';
import AsyncStorage from '@react-native-community/async-storage';

import _NetworkManager from './network-services/network-manager';
import _StorageService from './storage-service/storage-service';

const StorageService = new _StorageService(AsyncStorage);
const NetworkManager = new _NetworkManager(
  Config.AUTH_BASE,
  Config.AUTH_KEY,
  StorageService,
  fetch,
);
const FirestoreManager = new _NetworkManager(
  Config.FIRESTORE_BASE,
  Config.AUTH_KEY,
  StorageService,
  fetch,
);

export {NetworkManager, FirestoreManager, StorageService};
