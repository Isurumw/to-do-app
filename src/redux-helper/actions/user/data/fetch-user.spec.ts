import {FirestoreManager} from 'shared-services';
import fetchUser from './fetch-user';

jest.mock('shared-services/network-services/network-manager');
const mockedFirestoreManager =
  FirestoreManager as unknown as jest.Mocked<INetworkManager>;

const mockResponse = {
  documents: [
    {
      name: 'projects/todo-8013f/databases/(default)/documents/users/X00g6tPWHrFcYzpIynzJ',
      fields: {
        firstName: {
          stringValue: 'Isuru',
        },
        lastName: {
          stringValue: 'Madusanka',
        },
        phone: {
          stringValue: '+94 71 08 43196',
        },
        age: {
          integerValue: '29',
        },
        image: {
          stringValue:
            'https://firebasestorage.googleapis.com/v0/b/todo-8013f.appspot.com/o/isuru_profile.png?alt=media&token=8d3e0c01-f411-45d5-97f2-bbafcb53e897',
        },
        gender: {
          stringValue: 'male',
        },
      },
      createTime: '2021-09-27T09:35:54.418567Z',
      updateTime: '2021-09-27T18:22:03.888918Z',
    },
  ],
};

describe('check the behaviour of the fetch-user natwork service', () => {
  afterEach(() => {
    mockedFirestoreManager.request.mockClear();
  });

  it('fetch user successful', async () => {
    mockedFirestoreManager.request.mockImplementation(() => mockResponse);

    const response = await fetchUser();

    expect(mockedFirestoreManager.request).toBeCalledTimes(1);
    expect(response).toEqual({
      firstName: {
        stringValue: 'Isuru',
      },
      lastName: {
        stringValue: 'Madusanka',
      },
      phone: {
        stringValue: '+94 71 08 43196',
      },
      age: {
        integerValue: '29',
      },
      image: {
        stringValue:
          'https://firebasestorage.googleapis.com/v0/b/todo-8013f.appspot.com/o/isuru_profile.png?alt=media&token=8d3e0c01-f411-45d5-97f2-bbafcb53e897',
      },
      gender: {
        stringValue: 'male',
      },
    });
  });

  it('fetch user failed with an error', async () => {
    mockedFirestoreManager.request.mockImplementation(() => {
      throw new Error('Some API Error');
    });

    expect(fetchUser()).rejects.toThrowErrorMatchingSnapshot();
  });
});
