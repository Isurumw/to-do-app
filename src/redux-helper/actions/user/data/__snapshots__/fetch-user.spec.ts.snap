// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check the behaviour of the fetch-user natwork service fetch user failed with an error 1`] = `"Some API Error"`;
